### ISSUE TITLE
Fix Blizzard bug with TradeFrame

### ISSUE THREAD

**Issue by Anonomit** [tukui-org/ElvUI#277]
**Issue**
The TradeFrame does not always correctly show when someone has accepted the trade. The green highlihgting is sometimes missing.

**What I'd like**
The issue is a longstanding (years old) Blizzard bug, but could be patched by ElvUI. Whenever a player has accepted a trade, their section of the trade window should be highlighted green.

**Alternatives**
I made a WeakAura which fixes it, but it would be nice to have the fix built into ElvUI. I've suggested a method for fixing it below

**How to reproduce**
1. Open a trade window.
2. One player accepts trade.
3. Other player does not accept trade.
4. Enter in chat `/run GetItemInfo(25)`
5. Green highlighting should have disappeared from the window, obscuring the fact that a player has already accepted.
6. If the green highlighting still exists, then retry step 4, but replace 25 with other infrequently seen item IDs (ex. `2189, 3895, 5828`). You could also try hovering your cursor over items in the Inspect window, or in an item viewer addon like AtlasLoot.
7. Alternatively to step 4, try entering `/run TradeFrame_OnEvent(TradeFrame, "GET_ITEM_INFO_RECEIVED")`.

**Explanation**

`TradeFrame_SetAcceptState()` sets the highlighting. It is called on TRADE_ACCEPT_UPDATE and depends on arguments from this event. The actual accept state is never stored anywhere in the TradeFrame.
`TradeFrame_Update()` disables the highlighting. It is called on these events: TRADE_SHOW, TRADE_UPDATE, GET_ITEM_INFO_RECEIVED.

TRADE_SHOW and TRADE_UPDATE correctly hide the highlights, but GET_ITEM_INFO_RECEIVED is often (always?) irrelevant.

**Suggested Implementation**
ElvUI could internally track the trade accept state. The values would be set by hooking TRADE_ACCEPT_UPDATE, and would be cleared on TRADE_SHOW and TRADE_CLOSED. Hooking TradeFrame_Update and calling TradeFrame_SetAcceptState (passing the internal state) would restore the correct highlighting, which otherwise is lost on any GET_ITEM_INFO_RECEIVED.

**Mockup**

```lua
local TradeAcceptState = {}

local frame = CreateFrame"Frame"
frame:RegisterEvent"TRADE_SHOW"
frame:RegisterEvent"TRADE_CLOSED"
frame:RegisterEvent"TRADE_ACCEPT_UPDATE"
frame:SetScript("OnEvent", function(_, _, p, t)
  TradeAcceptState = {p or 0, t or 0}
end)

-- Could probably also achieve the same thing by listening for GET_ITEM_INFO_RECEIVED, but that's a busy event which happens often while the TradeFrame is hidden.
hooksecurefunc("TradeFrame_Update", function()
  TradeFrame_SetAcceptState(unpack(TradeAcceptState))
end)
```

**Example**
[Here is a fix in WeakAura format](https://wago.io/A6fqSc1Bd).

I have verified that my issue **DOES** exist in the default UI.

**Comment by kodewdle** [comment]
going to leave this here for right now, im not sure we will add it because any trade accepting stuff should be handled from blizzard. have you considered posting the issue to here:  https://github.com/Stanzilla/WoWUIBugs

**Comment by Anonomit** [comment]
I wasn't aware that this issue tracker existed. Thank you for pointing it out, and I've submitted [a report](https://github.com/Stanzilla/WoWUIBugs/issues/248) there.

I'll just point out that this doesn't affect the actual trade button, which does still enable/disable itself correctly. But I do understand the hesitancy with modifying anything TradeFrame-related, and it should be Blizzard's responsibility anyway. There was a Blizzard bug with MasterLoot last year that lasted awhile. Eventually I talked with Repooc in discord about it and ElvUI did patch that one temporarily. It was pretty critical though, it was causing all non-Blizzard dropdowns to break after clicking on a MasterLoot item.

Maybe someday Blizzard will give us an API call to actually check if a player has accepted a trade. :crossed_fingers:
